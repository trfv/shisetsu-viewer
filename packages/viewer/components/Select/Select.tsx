import MenuItem from "@mui/material/MenuItem";
import MuiSelect, { type SelectChangeEvent as MuiSelectChangeEvent } from "@mui/material/Select";
import type { FC } from "react";
import { box, type BoxSize } from "../Box";
import { SmallLabel } from "../Label";
import { Spacer } from "../Spacer";

export type SelectChangeEvent<T> = MuiSelectChangeEvent<T>;

type Props = {
  label: string;
  value: string;
  size?: BoxSize;
  onChange: (event: SelectChangeEvent<string>) => void;
  selectOptions: { value: string; label: string }[];
};

export const Select: FC<Props> = ({ label, value, size = "auto", onChange, selectOptions }) => {
  const Box = box(size);
  return (
    <Box component="label" display="flex" flexDirection="column">
      <SmallLabel label={label} />
      <Spacer axis="vertical" size={4} />
      <MuiSelect onChange={onChange} value={value} variant="standard">
        {selectOptions.map((option) => (
          <MenuItem key={option.value} value={option.value}>
            {option.label}
          </MenuItem>
        ))}
      </MuiSelect>
    </Box>
  );
};
