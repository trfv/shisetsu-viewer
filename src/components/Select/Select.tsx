import MenuItem from "@mui/material/MenuItem";
import MuiSelect, { SelectChangeEvent as MuiSelectChangeEvent } from "@mui/material/Select";
import type { FC } from "react";
import { box, BoxSize } from "../Box";
import { SmallLabel } from "../Label";
import { Spacer } from "../Spacer";

export type SelectChangeEvent<T> = MuiSelectChangeEvent<T>;

type Props = {
  label: string;
  value: string;
  size?: BoxSize;
  onChange: (event: SelectChangeEvent<string>) => void;
  selectOptions: { value: string; label: string }[];
};

export const Select: FC<Props> = ({ label, value, size = "auto", onChange, selectOptions }) => {
  const Box = box(size);
  return (
    <Box component="label" display="flex" flexDirection="column">
      <SmallLabel label={label} />
      <Spacer size={4} axis="vertical" />
      <MuiSelect value={value} onChange={onChange} variant="standard">
        {selectOptions.map((option) => (
          <MenuItem key={option.value} value={option.value}>
            {option.label}
          </MenuItem>
        ))}
      </MuiSelect>
    </Box>
  );
};
