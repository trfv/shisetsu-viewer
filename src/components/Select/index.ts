export { Select } from "./Select";
export type { SelectChangeEvent } from "./Select";
