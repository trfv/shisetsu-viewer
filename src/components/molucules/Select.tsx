import FormControl from "@material-ui/core/FormControl";
import MenuItem from "@material-ui/core/MenuItem";
import MuiSelect from "@material-ui/core/Select";
import React, { FC } from "react";
import FormLabel from "../atoms/FormLabel";

type SelectProps = {
  label: string;
  value: string;
  onChange: (event: React.ChangeEvent<{ value: unknown }>) => void;
  disabled?: boolean;
  selectOptions: { value: string; label: string }[];
};

const Select: FC<SelectProps> = ({
  label,
  value,
  onChange,
  disabled,
  selectOptions,
}: SelectProps) => {
  return (
    <FormControl>
      <FormLabel labelText={label} />
      <MuiSelect value={value} onChange={onChange} disabled={disabled}>
        {selectOptions.map((option, index) => (
          <MenuItem key={index} value={option.value}>
            {option.label}
          </MenuItem>
        ))}
      </MuiSelect>
    </FormControl>
  );
};

export default Select;
