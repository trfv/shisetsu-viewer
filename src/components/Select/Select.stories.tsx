import { useArgs } from "@storybook/preview-api";
import type { Meta, StoryObj } from "@storybook/react";
import { Select, SelectChangeEvent } from "./Select";

export default {
  component: Select,
} as Meta<typeof Select>;

export const Basic: StoryObj<typeof Select> = {
  args: {
    label: "label",
    selectOptions: [
      { value: "a", label: "A" },
      { value: "b", label: "B" },
      { value: "c", label: "C" },
    ],
  },
  argTypes: {
    value: {
      control: {
        type: "select",
        options: ["a", "b", "c"],
      },
    },
    selectOptions: {
      control: false,
    },
    size: {
      control: false,
    },
  },
  render: (args) => {
    const [{ value }, updateArgs] = useArgs();
    const onChange = (event: SelectChangeEvent<string>) =>
      updateArgs({ ...args, value: event.target.value });
    return <Select {...args} value={value ?? args.value} onChange={onChange} />;
  },
};
