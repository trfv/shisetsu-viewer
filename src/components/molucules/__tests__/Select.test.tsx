import "@testing-library/jest-dom/extend-expect";
import { render } from "@testing-library/react";
import React from "react";
import renderer from "react-test-renderer";
import Select from "../Select";

it("snapshot", () => {
  const props = {
    label: "select",
    value: "",
    onChange: jest.fn(),
    disabled: false,
    selectOptions: [
      {
        value: "value1",
        label: "label1",
      },
      {
        value: "value2",
        label: "label2",
      },
      {
        value: "value3",
        label: "label3",
      },
    ],
  };
  const component = renderer.create(<Select {...props} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it("interaction", () => {
  const props = {
    label: "select",
    value: "",
    onChange: jest.fn(),
    disabled: false,
    selectOptions: [
      {
        value: "value1",
        label: "label1",
      },
      {
        value: "value2",
        label: "label2",
      },
      {
        value: "value3",
        label: "label3",
      },
    ],
  };
  const select = render(<Select {...props} />);
  expect(select.queryByText("select")).toBeTruthy();
});
