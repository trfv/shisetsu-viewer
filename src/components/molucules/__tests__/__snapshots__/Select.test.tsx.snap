// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  className="MuiFormControl-root"
>
  <div
    className="MuiBox-root MuiBox-root-1"
  >
    select
  </div>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
    onClick={[Function]}
  >
    <div
      aria-haspopup="listbox"
      className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        dangerouslySetInnerHTML={
          Object {
            "__html": "&#8203;",
          }
        }
      />
    </div>
    <input
      aria-hidden={true}
      className="MuiSelect-nativeInput"
      onAnimationStart={[Function]}
      onChange={[Function]}
      required={false}
      tabIndex={-1}
      value=""
    />
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSelect-icon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
  </div>
</div>
`;
